### PR TITLE
Added validations for start and due date in Schedule tab inside  CCX coach dashboard

### PR DESCRIPTION
**Pull request by amir-qayyum-khan**
Hi 

In this PR 
- I am showing error message incase user do not specify a time with date.
- Showing error message in case start date is greater the due date i.e added date validation.

@pdpinch @giocalitri @pwilkins @jamiefolsom 

#### Note:
- If user do not specify both date and time at all then he can **Add item** in schedule tree. He can change dates from schedule tree any time just like in edX/master.

Issue https://github.com/mitocw/edx-platform/issues/79
issue https://github.com/mitocw/edx-platform/issues/39

screen shot 2015-10-07 at 3 09 16 pm

screen shot 2015-10-07 at 4 36 45 pm

### CODE REVIEW

**Comment by pdpinch** [comment]
@jamiefolsom you've been spending some time with the edX UI pattern library. Do they have anything in there for form validation errors?

**Comment by jamiefolsom** [comment]
Hi @pdpinch and @amir-qayyum-khan,

There is some guidance on client side form validation in this section: http://ux.edx.org/elements/forms/ 

I am not sure that pattern library assets are actually generally available throughout the platform as of yet however. In fact, clicking on the markup and styling tabs on that page, I don't see examples.

I will ask @talbs today.

**Comment by jamiefolsom** [comment]
Brian Talbot (@talbs, design lead at edx) tells me that the style rules are not in the platform yet. He says: 

> Right now its a cherry-pick, the good markup/match semantics will help you down the road and shouldn’t hurt much of the styling you’ll get out of the box for the LMS now so I’d try writing your markup against the UXPL’s way of forms and other elements and see what you get, then add specific styling.

So we might try cribbing from here: http://ux.edx.org/elements/forms/

File: lms/static/js/ccx/schedule.js [hunk 1]
```diff
@@ -104,23 +104,37 @@ var edx = edx || {};
...
+                var start = self.get_datetime('start');
+                var chapter = self.chapter_select.val(),
+                sequential = self.sequential_select.val(),
+                vertical = self.vertical_select.val(),
+                units = self.find_lineage(self.schedule,
+                chapter,
+                sequential === 'all' ? null : sequential,
+                vertical === 'all' ? null : vertical),
+                due = self.get_datetime('due');
+								if (self.valid_dates(start, due)) {
```
Comment (giocalitri) [inline lms/static/js/ccx/schedule.js]: please indent

**Comment by giocalitri** [comment]
I am not sure this is the right approach to do dates validation: I do not like the `try/catch` wrapping so many lines and in general I do not think `try/catch` is the right approach.
As @noisecapella also suggested, I would have the validation function return a list of errors and show an error if it is not empty.

I'm not sure what are the ex guidelines for javascript (or if they even exist), but I am not confident they will like something like this: honestly if it was python I would say 100% no.

cc @pdpinch

**Comment by amir-qayyum-khan** [comment]
@giocalitri I found try catch natural solution to this problem. I fixed issue with minimum change of code.

**Comment by pdpinch** [comment]
I've been spending more time with Studio and I'm noticing that the time widget in the courseware view defaults to 00:00, which sidesteps this whole issue. If the user deliberately clears the field, it still defaults to 00:00.  

Regarding the try/catch I have to agree with @giocalitri. There is **a lot** going on in this try block and any error will be caught and returned to the user as a problem with the time. 

Can we just go with 00:00 ?

**Comment by amir-qayyum-khan** [comment]
@pdpinch  and @giocalitri  please take a look.
Changed code logic plus removed ```try catch```

File: lms/static/js/ccx/schedule.js [hunk 2]
```diff
@@ -104,23 +104,36 @@ var edx = edx || {};
...
-	      units.map(self.show);
-	      var unit = units[units.length - 1];
-	      if (unit !== undefined && start) { unit.start = start; }
-	      if (unit !== undefined && due) { unit.due = due; }
-	      self.schedule_apply([unit], self.show);
-	      self.schedule_collection.set(self.schedule);
-	      self.dirty = true;
-	      self.render();
+             // Catching exception if date is without time and showing
+             // error message to user.
```
Comment (giocalitri) [inline lms/static/js/ccx/schedule.js]: I think this comment is outdated

File: lms/static/js/ccx/schedule.js [hunk 3]
```diff
@@ -104,23 +104,36 @@ var edx = edx || {};
...
+             var start = self.get_datetime('start');
+             var chapter = self.chapter_select.val(),
+             sequential = self.sequential_select.val(),
+             vertical = self.vertical_select.val(),
+             units = self.find_lineage(self.schedule,
+             chapter,
+             sequential === 'all' ? null : sequential,
+             vertical === 'all' ? null : vertical),
+             due = self.get_datetime('due');
```
Comment (giocalitri) [inline lms/static/js/ccx/schedule.js]: please indent this function call better

File: lms/static/js/ccx/schedule.js [hunk 4]
```diff
@@ -251,9 +264,27 @@ var edx = edx || {};
...
+          }
+
+          var dueDate = new Date(due);
+          var startDate = new Date(start);
+
+          if (dueDate < startDate) {
+            return [false, "Due date cannot be before start date."];
+          }
+          return [true, "ok"];
+	},
```
Comment (giocalitri) [inline lms/static/js/ccx/schedule.js]: why you don't just return a list of errors? if the `list.length` is `0`, the date is valid (so on line 119 it's easier to understand what's going on)

**Comment by amir-qayyum-khan** [comment]
Done @giocalitri

File: lms/static/js/ccx/schedule.js [hunk 5]
```diff
@@ -104,23 +104,35 @@ var edx = edx || {};
...
-	      self.render();
+	      sequential = self.sequential_select.val(),
+	      vertical = self.vertical_select.val(),
+	      units = self.find_lineage(self.schedule,
+	      chapter,
+	      sequential === 'all' ? null : sequential,
+	      vertical === 'all' ? null : vertical),
+	      due = self.get_datetime('due');
+	      var result = self.valid_dates(start, due);
+			  if (result.length === 0) {
```
Comment (giocalitri) [inline lms/static/js/ccx/schedule.js]: Please indent better

File: lms/static/js/ccx/schedule.js [hunk 6]
```diff
@@ -251,9 +263,29 @@ var edx = edx || {};
...
+          }
+
+          var dueDate = new Date(due);
+          var startDate = new Date(start);
+
+          if (dueDate < startDate) {
+            errorList[0] = "Due date cannot be before start date.";
+          }
+          return errorList;
+	},
```
Comment (giocalitri) [inline lms/static/js/ccx/schedule.js]: Question: is this function returning always only one value? If so you do not need to return a list. But I'm not sure about the logic you had in mind.

File: lms/static/js/ccx/schedule.js [hunk 7]
```diff
@@ -104,23 +104,35 @@ var edx = edx || {};
...
-	      self.schedule_apply([unit], self.show);
-	      self.schedule_collection.set(self.schedule);
-	      self.dirty = true;
-	      self.render();
+	      sequential = self.sequential_select.val(),
+	      vertical = self.vertical_select.val(),
+	      units = self.find_lineage(self.schedule,
+	      chapter,
+	      sequential === 'all' ? null : sequential,
+	      vertical === 'all' ? null : vertical),
```
Comment (pdpinch) [inline lms/static/js/ccx/schedule.js]: indentation

File: lms/static/js/ccx/schedule.js [hunk 8]
```diff
@@ -104,23 +104,37 @@ var edx = edx || {};
 	    // Add unit handler
 	    $('#add-unit-button').on('click', function(event) {
 	      event.preventDefault();
+	      // Default value of time is 00:00.
+	      var start = self.get_datetime('start');
```
Comment (justinabrahms) [inline lms/static/js/ccx/schedule.js]: If it's our style to use a single very large `var` block, then we should add this into that. ie: replace the `;` with a `,` and remove the var on line 109.

File: lms/static/js/ccx/schedule.js [hunk 9]
```diff
@@ -104,23 +104,37 @@ var edx = edx || {};
...
+	      sequential = self.sequential_select.val(),
+	      vertical = self.vertical_select.val(),
+	      units = self.find_lineage(
+	        self.schedule,
+	        chapter,
+	        sequential === 'all' ? null : sequential,
+	        vertical === 'all' ? null : vertical
+	      ),
+	      due = self.get_datetime('due');
+	      var errorMessage = self.valid_dates(start, due);
```
Comment (justinabrahms) [inline lms/static/js/ccx/schedule.js]: Same here. Move this var into the big var block just before it.

**Comment by amir-qayyum-khan** [comment]
@pdpinch i think this is ready for edX. Should i create PR there?

File: lms/static/js/ccx/schedule.js (hunk 8 above)
Comment (amir-qayyum-khan) [inline lms/static/js/ccx/schedule.js]: @justinabrahms I think this is ok to go check edX standards for javascript https://github.com/edx/edx-platform/wiki/Javascript-standards-for-the-edx-platform

File: lms/static/js/ccx/schedule.js (hunk 9 above)
Comment (amir-qayyum-khan) [inline lms/static/js/ccx/schedule.js]: @justinabrahms I think this is ok to go check edX standards for javascript https://github.com/edx/edx-platform/wiki/Javascript-standards-for-the-edx-platform

**Comment by amir-qayyum-khan** [comment]
@justinabrahms are you ok with my feedback? Need thumbs up from you.

**Comment by giocalitri** [comment]
This PR is fine with me when you agree with @justinabrahms on the style comments.

**Comment by justinabrahms** [comment]
:+1: on style.

**Comment by pdpinch** [comment]
@amir-qayyum-khan this can be submitted to edX. It's too late for this release, so you can do it tomorrow. 

@giocalitri can you give a thumbs up on the edX PR, once it's open?

**Comment by amir-qayyum-khan** [comment]
sure

**Comment by giocalitri** [comment]
> @giocalitri can you give a thumbs up on the edX PR, once it's open?

yep

**Comment by amir-qayyum-khan** [comment]
opened PR on edX platform https://github.com/edx/edx-platform/pull/10538

**Pull request closed without merging**